processing/src/main/java/io/druid/query/PrioritizedExecutorService.java

```diff
@@ -185,89 +190,215 @@ public int getQueueSize()
+  {
+    delegate.run();
+  }
 
-    @Override
-    public void run()
-    {
-      delegate.run();
-    }
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning)
+  {
+    return delegate.cancel(mayInterruptIfRunning);
+  }
 
-    @Override
-    public boolean cancel(boolean mayInterruptIfRunning)
-    {
-      return delegate.cancel(mayInterruptIfRunning);
-    }
+  @Override
+  public boolean isCancelled()
+  {
+    return delegate.isCancelled();
+  }
 
-    @Override
-    public boolean isCancelled()
-    {
-      return delegate.isCancelled();
-    }
+  @Override
+  public boolean isDone()
+  {
+    return delegate.isDone();
+  }
 
-    @Override
-    public boolean isDone()
-    {
-      return delegate.isDone();
-    }
+  @Override
+  public V get() throws InterruptedException, ExecutionException
+  {
+    return delegate.get();
+  }
 
-    @Override
-    public V get() throws InterruptedException, ExecutionException
-    {
-      return delegate.get();
-    }
+  @Override
+  public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
+  {
+    return delegate.get(timeout, unit);
+  }
 
-    @Override
-    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
-    {
-      return delegate.get(timeout, unit);
-    }
+  @Override
+  public void addListener(Runnable listener, Executor executor)
+  {
+    delegate.addListener(listener, executor);
+  }
 
-    @Override
-    public void addListener(Runnable listener, Executor executor)
-    {
-      delegate.addListener(listener, executor);
-    }
+  @Override
+  public int getPriority()
+  {
+    return priority;
+  }
+
+  @Override
+  public int compareTo(PrioritizedListenableFutureTask otherTask)
+  {
+    return Ints.compare(otherTask.getPriority(), getPriority());
+  }
+}
 
+class PrioritizedTaskBlockingQueue extends PriorityBlockingQueue<Runnable>
+{
+  private static final Comparator<Runnable> PRIORITY_AND_ENUMERATION_COMPARABLE = new Ordering<Runnable>()
+  {
     @Override
-    public int getPriority()
+    public int compare(@Nullable Runnable r1, @Nullable Runnable r2)
     {
-      return priority;
+      int compare = 0;
+      if (r1 instanceof Prioritized && r2 instanceof Prioritized) {
+        compare = Integer.compare(((Prioritized) r1).getPriority(), ((Prioritized) r2).getPriority());
+      }
+      if (compare == 0) {
+        if (r1 instanceof Enumerated && r2 instanceof Enumerated) {
+          compare = Long.compare(((Enumerated) r1).getEnumeration(), ((Enumerated) r2).getEnumeration());
+        }
+      }
+      return compare;
     }
+  }.reverse(); // Higher means more priority
+  final AtomicLong seqNum;
 
-    @Override
-    public int compareTo(PrioritizedListenableFutureTask otherTask)
-    {
-      return Ints.compare(otherTask.getPriority(), getPriority());
+  PrioritizedTaskBlockingQueue(ExecutorServiceConfig config, AtomicLong seqNum)
```

Comment:
I don't think you actually need to create a specialized priority queue for this. You can simply use the same priority from the existing code. 

Every item put into the queue is already an instance of `PrioritizedListenableFutureTask` which already implements `Comparable`. This is enforced by the `newTaskFor(Runnable)` and `newTaskFor(Callable)` in `PrioritizedExecutorService`.

All that's needed to make the queue FIFO, is to extend the `compareTo` method in `PrioritizedListenableFutureTask` to include a sequence number, similar to what is done in the example here: http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/PriorityBlockingQueue.html
Using a global static AtomicLong inside of the `PrioritizedListenableFutureTask` will remove the need to keep track of sequence numbers inside of the priority queue, and will still be fine for a very very long time, the same reasons you mentioned.

Once you do that you can get rid of both the special purpose priority queue as well as the enumerated interface.